File: app/metadata.py
import mutagen
from mutagen.easyid3 import EasyID3
from mutagen.flac import FLAC
from mutagen.id3 import ID3, APIC

import mutagen.flac
from spotify import SpotifySong
from utils import download_album_art


def add_metadata(file_name: str, song: SpotifySong, album_art_path: str):
    """
    Adds metadata to a file given it's path.
    File name must contain file extension as well.
    """

    # fix for invalid character in file name on windows
    if "?" in file_name:
        file_name = file_name.replace("?", "#")

    # writing textual metadata to the file
    try:
        meta = EasyID3(file_name)

    except mutagen.id3.ID3NoHeaderError:
        meta = mutagen.File(file_name, easy=True)
        meta.add_tags()

    meta["title"] = song.name
    meta["artist"] = "/".join(song.artists)
    meta["album"] = song.album_name
    meta["tracknumber"] = str(song.track_number)
    meta["discnumber"] = str(song.disc_number)

    # here we write the downloaded album art image to our downloaded song.
    meta = ID3(file_name)

    with open(album_art_path, "rb") as album_art:
        meta["APIC"] = APIC(
            # type 3 is for the front cover of a song
            encoding=3,
            mime="image/jpeg",
            type=3,
            data=album_art.read(),
        )

    # setting the v2 version to 3 is key- nothing will work without it
    meta.save(v1=2, v2_version=3)


def add_metadata_flac(file_name: str, song: SpotifySong):
    """
    Adds textual metadata to a FLAC file given it's path.
    File name must contain file extension as well.
    """

    meta = FLAC(file_name)

    try:
        meta.add_tags()

    except mutagen.flac.FLACVorbisError:
        # if this error occurs, it means that the vorbis comment block already
        # exists and we actually just need to start writing metadata
        meta.delete()

    meta["title"] = song.name
    meta["artist"] = "/".join(song.artists)
    meta["album"] = song.album_name
    meta["tracknumber"] = str(song.track_number)
    meta["discnumber"] = str(song.disc_number)

    # now, to write the album art to the file
    meta.add_picture()

    meta.save()


def metadata_controller(file_name: str, song: SpotifySong, dir: str, codec: str):
    """
    Handles the metadata writing process flow.
    """

    print("\nWriting metadata...")

    album_art_path = download_album_art(
        path=dir, link=song.cover_url, title=song.album_name
    )

    if codec == "flac" or codec == "vorbis":
        add_metadata_flac(file_name, song)

    else:
        add_metadata(file_name, song, album_art_path)

    print("\nFinished writing metadata!")


# if __name__ == "__main__":
# add_cover(file_path="./yASD-dl/Rauw Alejandro, Tainy - ¿Cuándo Fue#")

File: src/metadata.rs
use crate::spotify;

use std::fmt::Debug;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;

use lofty::{Accessor, Picture, Tag, TagExt, TaggedFileExt};
use log::error;

/// Adds metadata for the downloaded song given its path and requisite metadata information
pub fn add_metadata<P, S>(
    file_path: P,
    album_art_path: P,
    simple_song: spotify::SimpleSong,
    album_name: S,
) where
    P: AsRef<Path> + Debug,
    S: Into<String>,
{
    let error_msg = format!("Unable to write metadata for {} song!", simple_song.name);

    let mut tagged_file = match lofty::Probe::open(&file_path) {
        Err(err) => {
            println!("{error_msg}");
            error!(
                "error {} opening song file {:?} for writing metadata",
                err, file_path
            );
            return;
        }
        Ok(file) => match file.read() {
            Err(err) => {
                println!("{error_msg}");
                error!(
                    "error {} reading song file {:?} for writing metadata",
                    err, file_path
                );
                return;
            }
            Ok(tf) => tf,
        },
    };

    let tag = match tagged_file.primary_tag_mut() {
        Some(primary_tag) => primary_tag,
        None => {
            if let Some(first_tag) = tagged_file.first_tag_mut() {
                first_tag
            } else {
                // create, add and return a new tag
                let tag_type = tagged_file.primary_tag_type();

                tagged_file.insert_tag(Tag::new(tag_type));

                match tagged_file.primary_tag_mut() {
                    None => {
                        println!("{error_msg}");
                        error!("unable to add metadata tag to song file {:?}", file_path);
                        return;
                    }
                    Some(v) => v,
                }
            }
        }
    };

    let artists = simple_song.artists.join(", ");
    tag.set_artist(artists);
    tag.set_title(simple_song.name.clone());
    tag.set_album(album_name.into());
    tag.set_disk(simple_song.disc_number as u32);
    tag.set_track(simple_song.track_number);

    let album_art = match File::open(&album_art_path) {
        Err(err) => {
            println!("{error_msg}");
            error!("error {} reading album art file {:?}", err, album_art_path);
            return;
        }
        Ok(v) => v,
    };
    let mut reader = BufReader::new(album_art);

    let mut picture = match Picture::from_reader(&mut reader) {
        Err(err) => {
            println!("{error_msg}");
            error!("error {} reading album art file {:?}", err, album_art_path);
            return;
        }
        Ok(v) => v,
    };

    picture.set_pic_type(lofty::PictureType::CoverFront);
    tag.push_picture(picture);

    if let Err(err) = tag.save_to_path(file_path) {
        println!("{error_msg}");
        error!(
            "error {} saving file {:?} after writing metadata",
            err, album_art_path
        );
    }
}
